### ISSUE TITLE
Have a separate test and training window folder for training.

### ISSUE THREAD

**Issue by Tilps** [glinscott/leela-chess#595]
Currently every training run re-selects the test and training at random from a combined window.  This makes it very difficult to track persistent divergence as any prior divergence gets nullified by the new random selection.

**Comment by srt19170** [comment]
How about making the random number generator reproducible (by seeding, for example)?  This would also permit multiple reproducible runs by switching seeds.

**Comment by Tilps** [comment]
What is desired is for a game to not change whether it is in test or train when the window moves.  A seeded random number generator is not sufficient for that because the underlying data being shuffled has changed.
A seeded hash of the filename might be an option though - if the integer equivalent of the last 4 bytes of the hash is below train percentage of max int, then its train otherwise its test.
Seeding the random number generator wouldn't make training deterministic anyway - there are multiple file reader processes, so you get different interleaving depending on I/O latencies and process time allocations which are not predictable.

**Comment by mooskagh** [comment]
If testing and training set have consecutive moves, their evals may be also too similar so it's better to avoid.
I think we just need to separate training and test games early (something like `if (game_id % X = 0) then testing set`), and just keep them separately.

If we still want to separate positions not games, we can compute a hash (not seeded) of position and decide based on hash.

**Comment by trebe** [comment]
A related issue is that all games from a period are correlated / biased, since they are produced by the same ID and same search algo (although randomized).
So even if we fit perfectly (with no overfitting) to the window, with perfect validation error on that window, we might have lost the knowledge from previous windows.
So one idea: Have a window of training games, and different windows of validation, so that we can monitor the performance on each span of validation games, to not risk to fit only to the last batch of games

**Comment by Tilps** [comment]
It would be difficult to interpret those graphs from older validation games - the loss on them would be expected to increase (or sometimes decrease) because the net has gotten better, its gameplay is better, therefore its previous evaluations are technically 'wrong'.  The window slides, only changing a bit at a time, I think just being able to track the divergence across two consecutive training runs is the most important improvement here - otherwise its easier to think the divergence is actually just tensorboard noise.

Not to say its a bad idea, but possibly one which can just be performed on released nets, rather than integrated into the training flow.  Still, interpreting the results would be an art.

**Comment by trebe** [comment]
Sure, the loss on older validation games would be expected to be bigger than on more recent validation games. But not by *a lot*. So the monitoring could be used to verify that this loss doesn't sometimes completely diverge.
Also, my main point is that it should not be a goal by itself to have best possible validation error on a very recent window, since this might have the effect of erasing previous knowledge, which is no longer manifested in the most recent window.
So better to err to the side of less training than too much training. 
It is not necessarily bad to see the validation error go down by only a little, insted of "converging", during a training session.

**Comment by nousian** [comment]
I think idea of separating training games from tests on end of file hash is a good idea. This would clearly, easily and permanently split the games into two buckets, either training or testing. 

We could do this on ingestion to training server, and even mark games with .train.gz and .test.gz labels on their filename to easily identify them with scripting tools to easy debugging on server end. then you can just ls | grep | wc them easily from command line without hashing them on the fly.

**Comment by trebe** [comment]
I find it strange that we use the same game in multiple training sessions. Why not train for the same number of steps, but restricted to only new games.

But given that we do, here's an modification of the idea of this issue:
In each training period, use half of the _new games_ for validation, and the other half, together with the window games (including previous validation games) for training.
That way we will get a metric of the network performance on _purely new, never before seen games_, while still training on a window.
We can reuse previous validation games for training, but the other way around is not good.

**Comment by jjoshua2** [comment]
That seems like it has some nice advantages of using all games eventually
to reduce overfitting. But I'm not sure if it's worth it to only have test
games from one net. If you had test games from 2 or 3 net probably ok

**Comment by nousian** [comment]
Actual bucketing part in bash:

```
       TSTDIR="/ramdisk/data/test"
       TRNDIR="/ramdisk/data/train"
       THRESHOLD_VAL=6553

        sha=$(sha256sum /work/data/run1/$file | awk '{print $1}' | cut -c1-4)
        val=$((16#$sha))

        fn=$NETDIR/$file

        if [ $val -lt $THRESHOLD_VAL ]
        then
                        target=$TSTDIR/$file
        else
                        target=$TRNDIR/$file
        fi

       cp --preserve=timestamps $fn $target
```
